### ISSUE TITLE
Can't run in `/tmp` (but can in `/tmp/some-dir`)

### ISSUE THREAD

**Issue by Andrew15-5** [Enter-tainer/typst-preview#130]
**Describe the bug**

Running in `/tmp` gives an error:

```log
[2023-09-24T18:13:29Z ERROR typst_preview::actor::typst] TypstActor: failed to watch filesystem events
thread '<unnamed>' panicked at 'explicit panic', src/actor/typst.rs:119:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Which also means nothing will be re-rendered.

<details><summary>Full log</summary>

```log
[2023-09-24T18:40:16Z INFO  typst_preview] Arguments: CliArguments {
        font_paths: [],
        root: None,
        data_plane_host: "127.0.0.1:23625",
        control_plane_host: "127.0.0.1:23626",
        static_file_host: "127.0.0.1:23627",
        dont_open_in_browser: true,
        enable_partial_rendering: false,
        input: "a.typ",
    }
[2023-09-24T18:40:17Z INFO  typst_preview] Data plane server listening on: 127.0.0.1:23625
[2023-09-24T18:40:17Z INFO  typst_preview] Control plane server listening on: 127.0.0.1:23626
[2023-09-24T18:40:17Z INFO  typst_ts_compiler::service::diag] /a.typ: compiling ...
[2023-09-24T18:40:17Z INFO  typst_preview] Static file server listening on: 127.0.0.1:23627
[2023-09-24T18:40:17Z INFO  typst_ts_compiler::service::diag] /a.typ: Compilation succeeded in 330.052µs
[2023-09-24T18:40:17Z ERROR typst_preview::actor::typst] TypstActor: failed to watch filesystem events
thread '<unnamed>' panicked at 'explicit panic', src/actor/typst.rs:119:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-09-24T18:40:18Z INFO  typst_preview] Peer address: 127.0.0.1:46826
[2023-09-24T18:40:18Z INFO  typst_preview] New WebSocket connection: 127.0.0.1:46826
[2023-09-24T18:40:18Z DEBUG typst_preview::actor::render] RenderActor: waiting for message
[2023-09-24T18:40:18Z DEBUG typst_preview::actor::webview] WebviewActor: received message from websocket: Ok(Text("current"))
[2023-09-24T18:40:18Z DEBUG typst_preview::actor::render] RenderActor: has_full_render: true
[2023-09-24T18:40:18Z DEBUG typst_preview::actor::render] RenderActor: waiting for message
[2023-09-24T18:40:18Z DEBUG typst_preview::actor::webview] WebviewActor: received svg from renderer
```

</details>

**To Reproduce**
1. `cd /tmp`
2. `typst-preview file.typ`
3. See error

**Expected behavior**
No errors.

**Software version:**
typst-preview version: `v0.8.1` (https://github.com/Enter-tainer/typst-preview/commit/bf83af63e62515881b192387b6f0e552f9346c02)

**Comment by Myriad-Dreamin** [comment]
The `/tmp` directory is owned by root. You may have to run `sudo typst-preview /tmp/main.typ` If you want to do that.

**Comment by Andrew15-5** [comment]
Owning a directory means nothing if `o`thers can do whatever inside it (`rwx`):

```log
$ ls -ldh /tmp
drwxrwxrwt 49 root root 112K Sep 24 22:20 /tmp
```

I had a similar problem with https://github.com/neoclide/coc.nvim/issues/4538, but now it is resolved.

**Comment by Myriad-Dreamin** [comment]
It is not quite similar, typst-preview regards the `/tmp` as the root directory of your document file. And so that typst-preview cannot watch anything if it meets a access denied when it watches the root directory.

For some reason, we don't watch files separately but watch a single root directory. https://github.com/Enter-tainer/typst-preview/issues/114

**Comment by Andrew15-5** [comment]
Is the Node.js used as a JS runtime? Or is it some Rust server? Is there a way to watch a dir that can be read or written to? I mean, `inotifywait` can read `/tmp` just fine.

**Comment by Myriad-Dreamin** [comment]
We are using it, https://docs.rs/notify/latest/notify/. And I have printed the error just now, looks like it is complaining that it does not have permission to `/tmp/systemd-private-...`, some subdirectory of `/tmp`, instead of permission to `/tmp` itself.

**Comment by Andrew15-5** [comment]
Yes, it's exactly the same as in https://github.com/neoclide/coc.nvim/issues/4538.

Is the "reading the whole root dir" comes from typst-lsp or from the typst itself? Either way, it should be allowed to work in a single-file-only mode (with the support of local/public packages).

**Comment by Andrew15-5** [comment]
> We are using it, https://docs.rs/notify/latest/notify/.

If it's possible, we can just throw a warning about the fact that some dirs/files can't be accessed, but if all necessary files can be accessed, then it shouldn't matter anyway. You know, support more aggressive/unideal environment (root dir).

**Comment by Myriad-Dreamin** [comment]
> Is the "reading the whole root dir" comes from typst-lsp or from the typst itself?

This is from typst itself.

> it should be allowed to work in a single-file-only mode.

I think this is reasonable.

> You know, support more aggressive/unideal environment (root dir).

It won't be supported to watch entire system before we changing idea on #114. @Enter-tainer perhaps have some comment on it.

**Comment by Andrew15-5** [comment]
> support more aggressive/unideal environment (root dir).

https://github.com/typst/typst/issues/2227

**Comment by Andrew15-5** [comment]
> This is from typst itself.

No, typst itself (and typst-lsp) work fine in the `/tmp`: https://github.com/typst/typst/issues/2227#issuecomment-1733111711.

**Comment by Myriad-Dreamin** [comment]
We rejected it since watch files separately watching files will cause performance degrades, more resources overhead, and relative big code base to handle them.

Whatever, I have added a (optional) large patch to the typst.ts to make watch files separately work. There is 1.5k LoC code to come. Hopefully that it will work well.

I don't know which one is correct, but I do that.

Related typst.ts: [compiler/src/service/watch.rs in typst.ts](https://github.com/Myriad-Dreamin/typst.ts/blob/babff32490d75ccf65a6e7d31f6cffd57b576823/compiler/src/service/watch.rs)
Based on rust analyzer: [crates/vfs-notify/src/lib.rs in rust-analyzer](https://github.com/rust-lang/rust-analyzer/blob/d3cc3bc00e310ff49268ce0c593eaa6bf4724bbd/crates/vfs-notify/src/lib.rs)
